pkg/api/v1/clients/kube/resource_client_factory.go

```diff
@@ -64,137 +67,174 @@ type ResourceClientSharedInformerFactory struct {
-	sharedInformer := cache.NewSharedIndexInformer(
-		&cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				//if tweakListOptions != nil {
-				//	tweakListOptions(&options)
-				//}
-
-				if ctxWithTags, err := tag.New(ctx, tag.Insert(KeyOpKind, "list")); err == nil {
-					ctx = ctxWithTags
-				}
-				stats.Record(ctx, MLists.M(1), MInFlight.M(1))
-				defer stats.Record(ctx, MInFlight.M(-1))
-				return list(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (kubewatch.Interface, error) {
-				// if tweakListOptions != nil {
-				// 	tweakListOptions(&options)
-				// }
-
-				if ctxWithTags, err := tag.New(ctx, tag.Insert(KeyOpKind, "watch")); err == nil {
-					ctx = ctxWithTags
-				}
-
-				stats.Record(ctx, MWatches.M(1), MInFlight.M(1))
-				defer stats.Record(ctx, MInFlight.M(-1))
-				return watch(options)
+	informerType := reflect.TypeOf(rc.crd.Type)
+	namespaces := rc.namespaces // will always contain at least one element
+
+	resyncPeriod := f.defaultResync
+	if rc.resyncPeriod != 0 {
+		resyncPeriod = rc.resyncPeriod
+	}
+
+	// Create a shared informer for each of the given namespaces.
+	// NOTE: We do not distinguish between the value "" (all namespaces) and a regular namespace here.
+	for _, ns := range namespaces {
+
+		// To nip configuration errors in the bud, error if the registry already contains an informer for the given resource/namespace.
+		if forResourceType, exists := f.informers[informerType]; exists {
+			if _, exists := forResourceType[ns]; exists {
+				return errors.Errorf("Shared cache already contains informer for resource [%v] and namespace [%v]", informerType, ns)
+			}
+		}
+
+		if ctxWithTags, err := tag.New(ctx, tag.Insert(KeyNamespaceKind, ns)); err == nil {
+			ctx = ctxWithTags
+		}
+
+		list := rc.kube.ResourcesV1().Resources(ns).List
+		watch := rc.kube.ResourcesV1().Resources(ns).Watch
+		sharedInformer := cache.NewSharedIndexInformer(
+			&cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+					if ctxWithTags, err := tag.New(ctx, tag.Insert(KeyOpKind, "list")); err == nil {
+						ctx = ctxWithTags
+					}
+					stats.Record(ctx, MLists.M(1), MInFlight.M(1))
+					defer stats.Record(ctx, MInFlight.M(-1))
+					return list(options)
+				},
+				WatchFunc: func(options metav1.ListOptions) (kubewatch.Interface, error) {
+					if ctxWithTags, err := tag.New(ctx, tag.Insert(KeyOpKind, "watch")); err == nil {
+						ctx = ctxWithTags
+					}
+
+					stats.Record(ctx, MWatches.M(1), MInFlight.M(1))
+					defer stats.Record(ctx, MInFlight.M(-1))
+					return watch(options)
+				},
 			},
-		},
-		&v1.Resource{},  // TODO(yuval-k): can we make this rc.crd.Type ?
-		f.defaultResync, // TODO(yuval-k): make this configurable!
-		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
-	)
+			&v1.Resource{}, // TODO(yuval-k): can we make this rc.crd.Type ?
```

Comment:
@yuval-k Not without other changes. Tried just to see if it would work and got this error:
```
ERROR: logging before flag.Parse: E1231 15:42:13.213211    6035 reflector.go:315] github.com/solo-io/solo-kit/pkg/api/v1/clients/kube/controller/controller.go:88: expected type *v1.ChangeSet, but watch event object had type *v1.Resource
```
Thrown from [here](https://github.com/kubernetes/client-go/blob/7856fdbcc3be4d133402982f0a5069d251eb8459/tools/cache/reflector.go#L315).

Reply:
can you explain a bit more? is this still an issue?

Reply:
I was replying to the second `TODO` comment. The parameter is used by the reflector to check the type of the event object it receives from the `ListWatch`. By passing `&v1.Resource{}` it works, but if you pass `rc.crd.Type` you get the error I mentioned. 

I believe the parameter must match the resource type that the `WatchFunc` in line 121
```go
watch := rc.kube.ResourcesV1().Resources(ns).Watch
```
is defined for, if that makes sense. Yuval was probably wondering if by passing `rc.crd.Type` you could get type information with the event for free, but from what I was able to tell there is no way of getting anything more than our `*v1.Resource`.